Iban's Staff charge tracker

jerwuqu opened the pull request:
As suggested in #1408

Shows the remaining charges in Iban's Staff.

This is what it looks like in inventory:
![image](https://user-images.githubusercontent.com/3710677/38773684-2ac97f6e-4053-11e8-9317-5dc4f5b2a04d.png)

InfoBox counter when equipped:
![image](https://user-images.githubusercontent.com/3710677/38773694-55046d02-4053-11e8-9f34-b442d5ea0174.png)

There are settings for disabling both inventory overlay and the infobox.

The charges are counted using the cast animation and are reset upon using the staff on the well.

Heavily tested with Iban's Staff (u) and works well.

jerwuqu commented:
It looks like the JewelleryCount plugin tracks all charges based on item ID, which is very different from this implementation. I understand that charge tracking would be nicer to be kept together in the same plugin, but we already have similar features spread out elsewhere due to them working differently, such as the Binding Necklace. In my opinion both the code and feature would feel very out of place in JewelleryCount.

deathbeam commented:
Actually, even having the binding necklace in the jewellery plugin instead of the runecrafting plugin would make sense. Mainly when there is probably plan in future to show infoboxes for various jewellery with charges, what would be awesome to do from 1 place too.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/ibansstaff/IbansStaffPlugin.java:
```diff
@@ -0,0 +1,230 @@
+	@Subscribe
+	public void onItemContainerChanged(final ItemContainerChanged event)
+	{
+		if (!config.alwaysShow())
```
You should probably check if the event.getType is either inventory or equipment, because this will be quite a few checks everytime you open a bank.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/ibansstaff/IbansStaffConfig.java:
```diff
@@ -0,0 +1,66 @@
+		name = "Always show charges",
+		description = "Always displays amount of Iban's Staff charges"
+	)
+	default boolean alwaysShow()
```
This should probably not be an option, and just show the infobox when staff is equipped, otherwise show counter on the staff itself (it would be shown also when equipped i guess, see JewelleryCount). Just adding boolean for even showing the infobox should suffice.

jerwuqu replied:
Having the option be "Display infobox when equipped" and having a counter in the inventory sounds like the best solution. Will fix as soon as the discussed JewelleryCount changes have been made.

Adam- commented:
A plugin specifically for ban staff charges seems silly. I would try to make a more generic plugin that could track most other item charges too, like trident. Also there appears to be a bunch of unrelated clue stuff in this?

jerwuqu commented:
Yes, the plan is to merge it with the current JewelleryCount plugin after #1201 passes. It has clue related stuff because it depends on #1493.

jerwuqu commented:
Rebased against master for new `ItemContainerChanged` event by Adam.

Do I rename the plugin to something like "Weapon Charges" as a future plugin for Trident and friends, or do I wait for the renaming of JewelleryCount in the aforementioned PR and put it in that plugin?

Adam- commented:
Maybe we could just call this "weaponcharges"? I do not like #1201 since it is not just renaming the plugin but also seems to be logically rewriting most of it and also adding new features at once.

jerwuqu commented:
Renamed the plugin to Weapon Charges and changed the structure to easier allow for additional weapons to be added later.

It's hard to make assumptions about future additions to the code, but I did my best. 

Any feedback welcome.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesOverlay.java:
```diff
@@ -0,0 +1,94 @@
+		final Query equipmentQuery = new EquipmentItemQuery().slotEquals(WidgetInfo.EQUIPMENT_WEAPON);
+		final Collection<WidgetItem> itemSlots = new ArrayList<>();
+		itemSlots.addAll(Arrays.asList(queryRunner.runQuery(inventoryQuery)));
+		itemSlots.addAll(Arrays.asList(queryRunner.runQuery(equipmentQuery)));
```
I don't think creating so many arrays and collections here is good approach to do this

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesOverlay.java:
```diff
@@ -0,0 +1,94 @@
+			if (chargedWeapon != null)
+			{
+				final Integer charges = plugin.getCharges(chargedWeapon);
+				if (charges == null) continue;
```
Do not use oneliners like this for conditions

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesOverlay.java:
```diff
@@ -0,0 +1,94 @@
+				int chargesStringWidth = graphics.getFontMetrics().stringWidth(chargesString);
+
+				final Rectangle bounds = item.getCanvasBounds();
+				final TextComponent textComponent = new TextComponent();
```
Probably pre-create textComponent as instance field for class, and then just modify properties here

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,247 @@
+		return overlay;
+	}
+
+	public Integer getCharges(ChargedWeapon weapon)
```
Making all these package-private instead of public would be better

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,247 @@
+	{
+		if (counter != null)
+		{
+			if (counter.getChargedWeapon() == weapon) return;
```
Do not use one-liners for conditions

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,247 @@
+	{
+		for (ChargedWeapon weapon : ChargedWeapon.values())
+		{
+			if (itemId == weapon.getItemId()) return weapon;
```
Do not use one-liners for conditions

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,247 @@
+	@Subscribe
+	public void onItemContainerChanged(final ItemContainerChanged event)
+	{
+		processChargesCounter();
```
You could probably use the data of the event itself here (and in further methods)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,247 @@
+	@Subscribe
+	public void onChatMessage(ChatMessage event)
+	{
+		if (event.getType() != ChatMessageType.SERVER) return;
```
Do not use one-liners for conditions

jerwuqu commented:
Requested changes have been made

jerwuqu commented:
Rebased

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesOverlay.java:
```diff
@@ -0,0 +1,92 @@
+
+		final Query inventoryQuery = new InventoryWidgetItemQuery();
+		final Query equipmentQuery = new EquipmentItemQuery().slotEquals(WidgetInfo.EQUIPMENT_WEAPON);
+		WidgetItem[] itemSlots = ArrayUtils.addAll(queryRunner.runQuery(inventoryQuery),
```
Definitely avoid running queries each frame. At the very least, put them in a gametick event in the plugin. It might be better to hook to the item container change event instead.

jerwuqu replied:
Turns out running the query on the ItemContainerChanged event was more troublesome than expected since the WidgetItem:s are not updated when they change places on screen (due to being created when the actual query is ran). 

Every other plugin (slayer, runecraft, runepouch, etc) using `InventoryWidgetItemQuery` and `EquipmentItemQuery` runs a query on every render, so to me it seems reasonable for it to happen here as well. 

What do you think?

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesOverlay.java:
```diff
@@ -0,0 +1,92 @@
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		graphics.setFont(FontManager.getRunescapeSmallFont());
```
This entire overlay should probably be disable-able if you only want the infobox?

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/weaponcharges/WeaponChargesPlugin.java:
```diff
@@ -0,0 +1,276 @@
+			removeChargesCounter();
+		}
+
+		counter = new WeaponChargesCounter(itemManager.getImage(weapon.getItemId()), this, weapon);
```
Run with `-ea` and this should assert. You can't `getImage` outside of the client thread.

jerwuqu commented:
Rebased and ready for review
